### PR TITLE
feat(notification/teams): add teams notification

### LINKED ISSUE #17939: InfluxDB v2 - MS-Teams - Notification Endpoint
__Proposal:__
Add MS-Teams as Endpoint Notification.

__Current behavior:__
No exist this option currently

__Desired behavior:__
I want to enable notifications of alerts and send them to a MS-Teams group.

__Alternatives considered:__
Describe other solutions or features you considered.

__Use case:__
It does not exist in Cronograf. That be great have in InfluxDB v2.

I know there is some antipathy to MS Teams.
All the more i hope that influxdata jumps over its own shadow and manages to implement this feature request.

### PR DESCRIPTION
**Pull request by sranka**
Closes #17939

This PR adds support for Microsoft Teams in InfluxDB notification endpoints and notification rules. It depends on influxdata/flux#2866 that introduces a new *teams* package to send notifications.

What is in this PR
  * implementation and tests of teams endpoint and rule
  * swagger updated with teams notification rule and endpoint
  * updated UI to use the new swagger + new UI for teams endpoint and rule:
      * Create/Edit Telegram Notification Endpoint
image
      * Create/Edit Telegram Notification Rule
image
      * MS Teams Notification (displayed in https://messagecardplayground.azurewebsites.net/ )
image

What is not in this PR
  * teams endpoint and rule in =pkger= 
  * new UI tests

- [ ] [CHANGELOG.md](https://github.com/influxdata/influxdb/blob/master/CHANGELOG.md) updated with a link to the PR (not the Issue)
- [x] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [ ] Rebased/mergeable
- [x] Tests pass
- [x] http/swagger.yml updated (if modified Go structs or API)
- [ ] Documentation updated or issue created (provide link to issue/pr)
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

**Comment by sranka** [comment]
The PR could be merged after #18218, small conflicts are expected.

**Comment by sranka** [comment]
@glinton ... sure, can you please first
- take the telegram notifications back into master (I could possibly help with it, but I need some instructions or code pointers to know of how to add a feature flag part), I have a lot of of conflicts herein that are now hard to resolve. Teams and telegram notifications change the same files and this PR was done on top of telegram changes.
- approve #19082 , it fixes a major defect in the  management of endpoints

**Comment by glinton** [comment]
>take back or help with the (f67c3a1) telegram notifications into master

I'm not certain why that was reverted, so i don't want to un-revert it. My guess is that you'll need to split up that work into 2 prs also, one for the backend code, and one for the front for the same reasons as mentioned above. Reach out to those involved in the revert if you'd like to learn more.

>approve #19082

I (commented on the pr)[https://github.com/influxdata/influxdb/pull/19082#pullrequestreview-457090326], but can you add tests to show the behavior described and that the new code fixes that. otherwise it looks like a sane change

**Comment by sranka** [comment]
@glinton ... I have to wait herein for the telegram to be re-introduced back into master, the other way would be painful. I changed this PR to draft because of it.

**Comment by sranka** [comment]
@glinton ... this type of notification can be now enabled with a feature flag "notification-endpoint-teams", the implementation follows already approved #19135

File: notification/rule/teams.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+		return "", err
+	}
+	return ast.Format(p), nil
+}
+
+// GenerateFluxAST generates a flux AST for the teams notification rule.
+func (s *Teams) GenerateFluxAST(e *endpoint.Teams) (*ast.Package, error) {
+	f := flux.File(
+		s.Name,
+		flux.Imports("influxdata/influxdb/monitor", "contrib/sranka/teams", "influxdata/influxdb/secrets", "experimental"),
```
Comment (glinton) [inline notification/rule/teams.go:38]: not being familiar with how these imports work, is `contrib/sranka/teams` outdated since https://github.com/influxdata/flux/pull/2866 got merged?

File: notification/endpoint/teams.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+			Code: influxdb.EInvalid,
+			Msg:  "empty teams URL",
+		}
+	}
+	return nil
+}
+
+// MarshalJSON implement json.Marshaler interface.
+func (s Teams) MarshalJSON() ([]byte, error) {
+	type teamsAlias Teams
```
Comment (glinton) [inline notification/endpoint/teams.go:57]: nit: move this out of the function (similar to how you did in rule/teams)

File: notification/endpoint/teams.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+// Valid returns error if some configuration is invalid
+func (s Teams) Valid() error {
+	if err := s.Base.valid(); err != nil {
+		return err
+	}
+	if s.URL == "" {
+		return &influxdb.Error{
+			Code: influxdb.EInvalid,
+			Msg:  "empty teams URL",
```
Comment (glinton) [inline notification/endpoint/teams.go:49]: I'm not sure what the standard is for these messages, but make them uniform (begin with either a capital or lowercase on all)

File: notification/rule/teams.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+// Valid returns where the config is valid.
+func (s Teams) Valid() error {
+	if err := s.Base.valid(); err != nil {
+		return err
+	}
+	if s.Title == "" {
+		return &influxdb.Error{
+			Code: influxdb.EInvalid,
+			Msg:  "Teams Title is empty",
```
Comment (glinton) [inline notification/rule/teams.go:112]: if lowercase is preferred and `Teams` must be capitalized, perhaps reword this to `empty title for Teams`

File: notification/rule/teams.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+			Code: influxdb.EInvalid,
+			Msg:  "Teams MessageTemplate is empty",
+		}
+	}
+	return nil
+}
+
+// Type returns the type of the rule config.
+func (s Teams) Type() string {
+	return "teams"
```
Comment (glinton) [inline notification/rule/teams.go:126]:
```suggestion
	return endpoint.TeamsType
```

File: notification/rule/teams_test.go [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+								Key:   "baz",
+								Value: "bang",
+							},
+							Operator: influxdb.Equal,
+						},
+					},
+				},
+			},
+			script: `package main
+// foo
```
Comment (glinton) [inline notification/rule/teams_test.go:106]: nit: remove stray comment, unless it serves a purpose

File: notification/endpoint/endpoint_test.go [hunk 7]
```diff
@@ -288,6 +306,39 @@ func TestJSON(t *testing.T) {
...
+					ID:     influxTesting.MustIDBase16Ptr(id1),
+					Name:   "name1",
+					OrgID:  influxTesting.MustIDBase16Ptr(id3),
+					Status: influxdb.Active,
+					CRUDLog: influxdb.CRUDLog{
+						CreatedAt: timeGen1.Now(),
+						UpdatedAt: timeGen2.Now(),
+					},
+				},
+				URL: "https://outlook.office.com/webhook/0acbc9c2-c262-11ea-b3de-0242ac130004",
```
[the remark this answers is not in the document]
Comment (sranka) [reply notification/endpoint/endpoint_test.go:339]: It does not, serialization of a "real" endpoint is tested.

File: notification/rule/teams.go (hunk 1 above)
Comment (sranka) [reply notification/rule/teams.go:38]: https://github.com/influxdata/flux/tree/master/stdlib/contrib/sranka/teams is still the same, there are tests that make the this package stable. I also tested that the notification works manually using a new UI.

File: notification/endpoint/teams.go (hunk 3 above)
Comment (sranka) [reply notification/endpoint/teams.go:49]: lowercase is mostly used in the code, this one is OK

File: notification/rule/teams.go (hunk 4 above)
Comment (sranka) [reply notification/rule/teams.go:112]: changed to `teams: empty title`

File: notification/rule/teams_test.go (hunk 6 above)
Comment (sranka) [reply notification/rule/teams_test.go:106]: this comment is generated when formatting flux AST tree, it contains the name of the teams endpoint, which is `foo`

**Comment by sranka** [comment]
This PR cannot be approved and merged anymore, it contains both UI and server part that are in now different repositories. Two PRs are required now.
